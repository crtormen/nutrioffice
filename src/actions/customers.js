import database from '../firebase/firebase';

//Action generator ADD_CUSTOMER
export const addCustomer = (customer) => ({
    type: 'ADD_CUSTOMER',
    customer
});

//Pass a function instead of a object  
// export const startAddCustomer = (customerData = {}) => {
//     return (dispatch, getState) => { 
//         const uid = getState().auth.uid;
//         const { //default values
//             name = '', 
//             email = '',
//             phone = '', 
//             born = '',              
//             occupation = '', 
//             instagram = ''
//         } = customerData;
//         const customer = { name, email, phone, born, occupation, instagram };

//         return database.ref(`users/${uid}/customers`).push(customer).then((ref) => { 
//             dispatch(addCustomer({
//                 id: ref.key,
//                 ...customer
//             }));
//         });
//     };
// };

//Action generator for EDIT_CUSTOMER
export const editCustomer = (id, updates) => ({
    type: 'EDIT_CUSTOMER',
    id,
    updates
});

// export const startEditExpense = (id, updates) => {
//     return (dispatch, getState) => {
//         const uid = getState().auth.uid;
//         return database.ref(`users/${uid}/expenses/${id}`).update(updates).then(() => {
//             dispatch(editExpense(id, updates));
//         })
//     };
// };

//Action generator for REMOVE_CUSTOMER
export const removeCustomer = ({ id } = {}) => ({
    type: 'REMOVE_CUSTOMER',
    id
});

// export const startRemoveExpense = ({ id } = {}) => {
//     return (dispatch, getState) => {
//         const uid = getState().auth.uid;
//         return database.ref(`users/${uid}/expenses/${id}`).remove().then(() => {
//             dispatch(removeExpense({ id }));
//         });
//     };
// };


//Action generator for SET_CUSTOMERS
export const setCustomers = (customers) => ({
    type: 'SET_CUSTOMERS',
    customers
});

//Action Generator START_SET_EXPENSES
// export const startSetExpenses = () => {
//     return (dispatch, getState) => {
//         const uid = getState().auth.uid;
//         //returns a promises, so that when using startSetExpenses, we can use "then"
//         return database.ref(`users/${uid}/expenses`).once('value').then((snapshot) => {
//             let expenses = [];
//             snapshot.forEach((childSnapshot) => {
//                 expenses.push({
//                     id: childSnapshot.key,
//                     ...childSnapshot.val()
//                 })
//             });

//             dispatch(setExpenses(expenses));
//         });        
//     };
// };
