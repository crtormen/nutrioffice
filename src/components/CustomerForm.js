import React from 'react';


export default class CustomerForm extends React.Component {
    constructor(props) {
        super(props);

        this.state = {
            name: props.customer ? props.customer.name : '',
            email: props.customer ? props.customer.email : '',
            phone: props.customer ? props.customer.phone : '',
            born: props.customer ? props.customer.born : '',
            occupation: props.customer ? props.customer.occupation : '',
            instagram: props.customer ? props.customer.instagram : '',
           // calendarFocused: false,
            error: ''
        };
    }


    onSubmit = (e) => {
        e.preventDefault();

        if (!this.state.name || !this.state.email || !this.state.phone) {
            this.setState(() => ({ error: "VocÃª deve informar seu nome, email e celular."})); 
        } else {
            this.setState(() => ({ error: ""})); 
            // Runs parent method
            this.props.onSubmit({
                name: this.state.name,
                email: this.state.email,
                phone: this.state.phone,
                born: this.state.born,
                occupation: this.state.occupation,
                instagram: this.state.instagram                 
            });
        };

    }

    onNameChange = (e) => {
        const name = e.target.value;
        this.setState(() => ({ name }));
    };

    onEmailChange = (e) => {
        const email = e.target.value;
        this.setState(() => ({ email }));
    };

    onPhoneChange = (e) => {
        const phone = e.target.value;
        this.setState(() => ({ phone }));
    };

    onOccupationChange = (e) => {
        const occupation = e.target.value;
        this.setState(() => ({ occupation }));
    };

    onBornChange = (e) => {
        const born = e.target.value;
        this.setState(() => ({ born }));
    };

    onInstagramChange = (e) => {
        const instagram = e.target.value;
        this.setState(() => ({ instagram }));
    };

    render() {
        return (
            <div>
            {this.state.error && <p>{this.state.error}</p>}
            <form onSubmit={this.onSubmit}>
                    <input 
                        type="text" 
                        placeholder="Nome completo" 
                        autoFocus 
                        value={this.state.name}
                        onChange={this.onNameChange}
                    />
                    <input 
                        type="text" 
                        placeholder="Data de nascimento" 
                        value={this.state.born}
                        onChange={this.onBornChange}
                    />
                    <input 
                        type="text" 
                        placeholder="Email"
                        value={this.state.email}
                        onChange={this.onEmailChange}
                    />
                    <input 
                        type="text" 
                        placeholder="Celular" 
                        value={this.state.phone}
                        onChange={this.onPhoneChange}
                    />
                    <input 
                        type="text" 
                        placeholder="ProfissÃ£o" 
                        value={this.state.occupation}
                        onChange={this.onOccupationChange}
                    />
                    <input 
                        type="text" 
                        placeholder="instagram" 
                        value={this.state.instagram}
                        onChange={this.onInstagramChange}
                    />
                <button>Salvar Dados</button>
            </form>
            </div>
        );
    }
}
