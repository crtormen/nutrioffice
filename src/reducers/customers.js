const customerReducerDefaultState = [];

export default (state = customerReducerDefaultState, action) => {
    switch (action.type) {
        case 'ADD_CUSTOMER':
            return [...state, action.customer];
        case 'EDIT_CUSTOMER':
            return state.map((customer) => {
                if (customer.id === action.id) {
                    return {
                        ...customer,
                        ...action.updates //create new object based ond the previous, with updates and concatenations
                    }
                }
                return customer; //do nothing if not found
            });
        case 'REMOVE_CUSTOMER':
            return state.filter(({ id }) => id !== action.id ); ////return all in a new array, except that one where filter not match 
        case 'SET_CUSTOMERS':
            return action.customers;
        default: 
            return state;
    };
};