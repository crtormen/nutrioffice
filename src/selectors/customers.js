export default (customers) => {
    return customers;
}



// Get visible expenses, filtering by text, period of time, and allows to sort by date and amount
// export default (expenses, { text, sortBy, startDate, endDate }) => { 
//     return expenses.filter((expense) => {
//         const startDateMatch = startDate ? startDate.isSameOrBefore(moment(expense.createdAt), 'day') : true;
//         const endDateMatch = endDate ? endDate.isSameOrAfter(moment(expense.createdAt), 'day') : true;
//         const textMatch = expense.description.toLowerCase().includes(text.toLowerCase());

//         return startDateMatch && endDateMatch && textMatch;
//     }).sort((a, b) => {  //highest first
//         if (sortBy === 'date')
//             return a.createdAt < b.createdAt ? 1 : -1;
//         if (sortBy === 'amount')
//             return a.amount < b.amount ? 1 : -1;

//     });
// };
