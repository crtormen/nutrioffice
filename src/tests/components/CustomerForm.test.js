import React from 'react';
import { shallow } from 'enzyme';
import CustomerForm from '../../components/CustomerForm';
import { customers } from '../fixtures/customers';

test('should render CustomerForm correctly', () => {
    const wrapper = shallow(<CustomerForm />);
    expect(wrapper).toMatchSnapshot();
});

test('should render CustomerForm with customer data', () => {
    const wrapper = shallow(<CustomerForm customer={customers[0]} />)
    expect(wrapper).toMatchSnapshot();
});

test('should render error for invalid form submission', () => {
    const wrapper = shallow(<CustomerForm />);
    expect(wrapper).toMatchSnapshot();
    wrapper.find('form').simulate('submit', {
        preventDefault: () => { }
    });
    expect(wrapper.state('error').length).toBeGreaterThan(0);
    expect(wrapper).toMatchSnapshot();
});

test("should set name on input change", () => {
    const value = "New name";
    const wrapper = shallow(<CustomerForm />);
    wrapper.find('input').at(0).simulate('change', {
        target: { value }
    });
    expect(wrapper.state('name')).toBe(value);
});

test("should set born on input change", () => {
    const value = "10/10/10";
    const wrapper = shallow(<CustomerForm />);
    wrapper.find('input').at(1).simulate('change', {
        target: { value }
    });
    expect(wrapper.state('born')).toBe(value);
});

test("should set email on valid input", () => {
    const value = "Newemail@email.com";
    const wrapper = shallow(<CustomerForm />);
    wrapper.find('input').at(2).simulate('change', {
        target: { value }
    });
    expect(wrapper.state('email')).toBe(value);
});

// test("should set email on invalid input", () => {
//     const value = "Newemail.email.com";
//     const wrapper = shallow(<CustomerForm />);
//     wrapper.find('input').at(2).simulate('change', {
//         target: { value }
//     });
//     expect(wrapper.state('email')).toBe('');
// });

test("should set phone on input change", () => {
    const value = "8888888";
    const wrapper = shallow(<CustomerForm />);
    wrapper.find('input').at(3).simulate('change', {
        target: { value }
    });
    expect(wrapper.state('phone')).toBe(value);
});

test("should set occupation on input change", () => {
    const value = "New name";
    const wrapper = shallow(<CustomerForm />);
    wrapper.find('input').at(4).simulate('change', {
        target: { value }
    });
    expect(wrapper.state('occupation')).toBe(value);
});

test("should set instagram on input change", () => {
    const value = "New name";
    const wrapper = shallow(<CustomerForm />);
    wrapper.find('input').at(5).simulate('change', {
        target: { value }
    });
    expect(wrapper.state('instagram')).toBe(value);
});