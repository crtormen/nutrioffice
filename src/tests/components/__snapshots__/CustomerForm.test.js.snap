// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render CustomerForm correctly 1`] = `
<div>
  <form
    onSubmit={[Function]}
  >
    <input
      autoFocus={true}
      onChange={[Function]}
      placeholder="Nome completo"
      type="text"
      value=""
    />
    <input
      onChange={[Function]}
      placeholder="Data de nascimento"
      type="text"
      value=""
    />
    <input
      onChange={[Function]}
      placeholder="Email"
      type="text"
      value=""
    />
    <input
      onChange={[Function]}
      placeholder="Celular"
      type="text"
      value=""
    />
    <input
      onChange={[Function]}
      placeholder="Profissão"
      type="text"
      value=""
    />
    <input
      onChange={[Function]}
      placeholder="instagram"
      type="text"
      value=""
    />
    <button>
      Salvar Dados
    </button>
  </form>
</div>
`;

exports[`should render CustomerForm with customer data 1`] = `
<div>
  <form
    onSubmit={[Function]}
  >
    <input
      autoFocus={true}
      onChange={[Function]}
      placeholder="Nome completo"
      type="text"
      value="Claudio"
    />
    <input
      onChange={[Function]}
      placeholder="Data de nascimento"
      type="text"
      value="02/03/1987"
    />
    <input
      onChange={[Function]}
      placeholder="Email"
      type="text"
      value="crtormen@gmail.com"
    />
    <input
      onChange={[Function]}
      placeholder="Celular"
      type="text"
      value="51993464776"
    />
    <input
      onChange={[Function]}
      placeholder="Profissão"
      type="text"
      value="software engineer"
    />
    <input
      onChange={[Function]}
      placeholder="instagram"
      type="text"
      value="crtormen"
    />
    <button>
      Salvar Dados
    </button>
  </form>
</div>
`;

exports[`should render error for invalid form submission 1`] = `
<div>
  <form
    onSubmit={[Function]}
  >
    <input
      autoFocus={true}
      onChange={[Function]}
      placeholder="Nome completo"
      type="text"
      value=""
    />
    <input
      onChange={[Function]}
      placeholder="Data de nascimento"
      type="text"
      value=""
    />
    <input
      onChange={[Function]}
      placeholder="Email"
      type="text"
      value=""
    />
    <input
      onChange={[Function]}
      placeholder="Celular"
      type="text"
      value=""
    />
    <input
      onChange={[Function]}
      placeholder="Profissão"
      type="text"
      value=""
    />
    <input
      onChange={[Function]}
      placeholder="instagram"
      type="text"
      value=""
    />
    <button>
      Salvar Dados
    </button>
  </form>
</div>
`;

exports[`should render error for invalid form submission 2`] = `
<div>
  <p>
    Você deve informar seu nome, email e celular.
  </p>
  <form
    onSubmit={[Function]}
  >
    <input
      autoFocus={true}
      onChange={[Function]}
      placeholder="Nome completo"
      type="text"
      value=""
    />
    <input
      onChange={[Function]}
      placeholder="Data de nascimento"
      type="text"
      value=""
    />
    <input
      onChange={[Function]}
      placeholder="Email"
      type="text"
      value=""
    />
    <input
      onChange={[Function]}
      placeholder="Celular"
      type="text"
      value=""
    />
    <input
      onChange={[Function]}
      placeholder="Profissão"
      type="text"
      value=""
    />
    <input
      onChange={[Function]}
      placeholder="instagram"
      type="text"
      value=""
    />
    <button>
      Salvar Dados
    </button>
  </form>
</div>
`;
